File: sort/py_impl/shell_sort.py
#!/usr/bin/env python3
# coding: utf-8
#
# author: kaku
# date: 19/10/12

# GitHub:
#
#   https://github.com/kakukosaku
#
# © 2019-2022 Kaku Kosaku All Rights Reserved

from typing import List, NoReturn


def shell_sort(arr: List[int], array_size: int) -> NoReturn:
    increment = array_size // 2
    while increment >= 1:

        i = increment
        while i < array_size:
            # from small to large
            if arr[i] < arr[i - increment]:
                tmp = arr[i]
                j = i - increment
                while j >= 0 and tmp < arr[j]:
                    arr[j + increment] = arr[j]
                    j -= increment

                arr[j + increment] = tmp

            i += 1

        increment = increment // 2

File: sort/c_impl/shell_sort.c
//
// Created by kaku on 2019/10/13.
//

#include "array.h"

void shell_sort(int *arr, int array_size) {
    int i, j, temp;
    for (int increment = array_size / 2; increment > 0; increment = increment / 2) {
        for (i = increment; i < array_size; i++) {
            temp = arr[i];
            if (arr[i] < arr[i - increment]) {
                for (j = i - increment; j >= 0 && arr[j] > temp; j -= increment) {
                    arr[j + increment] = arr[j];
                }
                arr[j + increment] = temp;
            }
        }
    }
}
